File: 3121-count-the-number-of-special-characters-ii/3121-count-the-number-of-special-characters-ii.py
class Solution:
    def numberOfSpecialChars(self, word: str) -> int:
        last_lower = {}
        first_upper = {}
        invalid = set()
        for i, ch in enumerate(word):
            letter = ch.lower()
            if ch.islower():
                last_lower[letter] = i
                if letter in first_upper:
                    invalid.add(letter)
            else:
                if letter not in first_upper:
                    first_upper[letter] = i
        special_count = 0
        for letter in last_lower:
            if letter in first_upper and letter not in invalid:
                special_count += 1
        return special_count

File: 3121-count-the-number-of-special-characters-ii/3121-count-the-number-of-special-characters-ii.java
class Solution {
    public int numberOfSpecialChars(String word) {
        int[] lastLower = new int[26];
        int[] firstUpper = new int[26];
        Arrays.fill(lastLower, -1);
        Arrays.fill(firstUpper, -1);
        Set<Integer> invalid = new HashSet<>();
        for (int i = 0; i < word.length(); i++) {
            char ch = word.charAt(i);
            if (ch >= 'a' && ch <= 'z') {
                int idx = ch - 'a';
                lastLower[idx] = i;
                if (firstUpper[idx] != -1) {
                    invalid.add(idx);
                }
            } else {
                int idx = ch - 'A';
                if (firstUpper[idx] == -1) {
                    firstUpper[idx] = i;
                }
            }
        }
        int specialCount = 0;
        for (int i = 0; i < 26; i++) {
            if (lastLower[i] != -1 &&
                firstUpper[i] != -1 &&
                !invalid.contains(i)) {
                specialCount++;
            }
        }
        return specialCount;
    }
}